Libraries/Opc.Ua.Security.Certificates/Extensions/X509AuthorityKeyIdentifierExtension.cs

```diff
@@ -0,0 +1,340 @@
+                        buffer.Append(", ");
+                    }
+                }
+
+                buffer.Append(kSerialNumber);
+                buffer.Append("=");
+                buffer.Append(m_serialNumber.ToHexString(true));
+            }
+            return buffer.ToString();
+        }
+
+        /// <summary>
+        /// Initializes the extension from ASN.1 encoded data.
+        /// </summary>
+        public override void CopyFrom(AsnEncodedData asnEncodedData)
+        {
+            if (asnEncodedData == null) throw new ArgumentNullException(nameof(asnEncodedData));
+            base.Oid = asnEncodedData.Oid;
+            base.RawData = asnEncodedData.RawData;
+            Decode(asnEncodedData.RawData);
+        }
+        #endregion
+
+        #region Public Properties
+        /// <summary>
+        /// The OID for a Authority Key Identifier extension.
+        /// </summary>
+        public const string AuthorityKeyIdentifierOid = "2.5.29.1";
+
+        /// <summary>
+        /// The alternate OID for a Authority Key Identifier extension.
+        /// </summary>
+        public const string AuthorityKeyIdentifier2Oid = "2.5.29.35";
+
+        /// <summary>
+        /// The identifier for the key as a little endian hexadecimal string.
+        /// </summary>
+        public string KeyIdentifier => m_keyIdentifier.ToHexString();
+
+        /// <summary>
+        /// The identifier for the key as a byte array.
+        /// </summary>
+        public byte[] GetKeyIdentifier() => m_keyIdentifier;
+
+        /// <summary>
+        /// A list of distinguished names for the issuer.
+        /// </summary>
+        public X500DistinguishedName Issuer => m_issuer;
+
+        /// <summary>
+        /// The serial number of the authority key as a big endian hexadecimal string.
+        /// </summary>
+        public string SerialNumber => m_serialNumber.ToHexString(true);
+
+        /// <summary>
+        /// The serial number of the authority key as a byte array in little endian order.
+        /// </summary>
+        public byte[] GetSerialNumber() => m_serialNumber;
+        #endregion
+
+        #region Private Methods
+        private byte[] Encode()
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+            writer.PushSequence();
+
+            if (m_keyIdentifier != null)
+            {
+                Asn1Tag keyIdTag = new Asn1Tag(TagClass.ContextSpecific, 0);
+                writer.WriteOctetString(m_keyIdentifier, keyIdTag);
+            }
+
+            if (m_issuer != null)
+            {
+                Asn1Tag issuerNameTag = new Asn1Tag(TagClass.ContextSpecific, 1);
+                writer.PushSequence(issuerNameTag);
+
+                // Add the issuer to constructed context-specific 4 (GeneralName.directoryName)
+                Asn1Tag directoryNameTag = new Asn1Tag(TagClass.ContextSpecific, 4, true);
+                writer.PushSetOf(directoryNameTag);
```

Comment:
This is really an explicit tag wrapper required by the target type being a choice.  Sequence is what I normally use for that, since set-ofs are weird (and PopSetOf will run a sorting algorithm that sorts the one thing).

X.680 2015-08 31.2.7: "The tagging construction specifies explicit tagging if any of the following holds: ... (c) ... the type defined by "Type" is an untagged choice type, ... "

Reply:
TODO

Reply:
added note for future